### PR TITLE
Settings menu

### PR DESCRIPTION
**Pull request by twjvdhorst**
Done with the settings menu. Only thing missing is the fow button, but I can't get that to work

### CODE REVIEW

**Comment by jorisBarkema** [comment]
Would be better if instead of going back by pressing p, as it is now, you can press p to enter the settings menu from anywhere, and close it to return to where you called it from, but I think that last bit could be hard, not sure exactly how the UI works. At least press p to enter settings menu instead of just going back to menu should be doable, and I think that would be an improvement.

Custom resolution is very fancy though and volume slider is great!

File: MeesGame/MeesGame.cs [hunk 1]
```diff
@@ -28,13 +28,14 @@ protected override void LoadContent()
         {
             base.LoadContent();
             screen = new Point(1440, 825);
-            windowSize = new Point(1024, 586);
+            //windowSize = new Point(1024, 586);
```
Comment (jorisBarkema) [inline MeesGame/MeesGame.cs]: this comment seems redundant

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 2]
```diff
@@ -10,8 +10,19 @@ public class TileField : GameObjectGrid, IDiscreteField, ITileField
         private static readonly Point NO_POINT = new Point(-1, -1);
         private static readonly Point DEFAULT_START = new Point(1, 1);
         private Point start = NO_POINT;
-        private int sightRadius = 3;
 
+        public Point Start
+        {
+            get
+            {
```
Comment (jorisBarkema) [inline MeesGame/Gameplay/GameObjects/TileField.cs]: Don;t think you need to change anything in tilefied

**Comment by jorisBarkema** [comment]
I do not have enough confidence in myself to review the rest of this.

File: MeesGame/Gameplay/GameObjects/TileField.cs (hunk 2 above)
Comment (LisanneW) [reply MeesGame/Gameplay/GameObjects/TileField.cs]: Even makes the game crash if you try to run it.

Comment (LisanneW) [reply MeesGame/Gameplay/GameObjects/TileField.cs]: I think something went wrong while merging the master branch because this property was relocated

**Comment by jensj12** [comment]
Since the resolution setting causes a few bugs, I think it's better to leave it out and just fix fullscreen mode.

**Comment by LisanneW** [comment]
@nessemis or @twjvdhorst can you fix the issues with the resolution settings or put that in another pull request so this can be merged?

File: MeesGame/Gameplay/GameManagement/GameEnvironment.cs [hunk 3]
```diff
@@ -119,9 +125,9 @@ protected override void LoadContent()
     protected void HandleInput()
     {
         inputHelper.Update();
-        if (inputHelper.KeyPressed(Keys.Escape))
+        if (inputHelper.KeyPressed(Keys.P))
         {
-            Exit();
+            gameStateManager.SwitchTo("SettingsMenuState");
```
Comment (jensj12) [inline MeesGame/Gameplay/GameManagement/GameEnvironment.cs]: Duplicate code. This is already in the gamestatemanager.

File: MeesGame/Gameplay/GameManagement/GameStateManager.cs [hunk 4]
```diff
@@ -53,7 +53,7 @@ public string PreviousGameState
 
     public void HandleInput(InputHelper inputHelper)
     {
-        if (inputHelper.KeyPressed(Keys.P)) SwitchTo("TitleMenuState");
+        if (inputHelper.KeyPressed(Keys.P)) SwitchTo("SettingsMenuState");
```
Comment (jensj12) [inline MeesGame/Gameplay/GameManagement/GameStateManager.cs]: Add escape here. Either make it go to the settings menu or main menu.

File: MeesGame/Gameplay/GameObjects/TileField.cs (hunk 2 above)
Comment (jensj12) [reply MeesGame/Gameplay/GameObjects/TileField.cs]: Move the property back to where it was.

File: MeesGame/Gameplay/GameStates/LoadMenuState.cs [hunk 5]
```diff
@@ -64,10 +65,18 @@ public LoadMenuState()
                 BeginLevel();
             });
 
+            editorButton = new SpriteSheetButton(new CenteredLocation(0, 25, true, false), null, Strings.map_editor, (UIComponent o) =>
+            {
+                //reset the gamestate to open blank;
+                GameEnvironment.GameStateManager.GetGameState("EditorState").Reset();
+                GameEnvironment.GameStateManager.SwitchTo("EditorState");
```
Comment (jensj12) [inline MeesGame/Gameplay/GameStates/LoadMenuState.cs:70]: First switch to the new game state, then reset the current game state.

File: MeesGame/UI/Components/Slider.cs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                    Dragging = false;
+                    return;
+                }
+
+                Value = startValue + (int)((inputHelper.MousePosition.X - startMouseX) * (max - min) / SliderBarDimensions.X);
+
+                if (Value > max)
+                    Value = max;
+                if (Value < min)
+                    Value = min;
```
Comment (jensj12) [inline MeesGame/UI/Components/Slider.cs]: Use MathHelper.Clamp(Value,min,max)

File: MeesGame/UI/DefaultUIValues.settings [hunk 7]
```diff
@@ -41,11 +41,18 @@
...
+    </Setting>
+    <Setting Name="SliderColor" Type="System.Drawing.Color" Scope="Application">
+      <Value Profile="(Default)">White</Value>
+	</Setting>
     <Setting Name="RadioButtonDimen1" Type="System.Int32" Scope="Application">
       <Value Profile="(Default)">20</Value>
     </Setting>
     <Setting Name="RadioButtonDimen" Type="System.Int32" Scope="Application">
       <Value Profile="(Default)">20</Value>
+>>>>>>> refs/remotes/origin/master
```
Comment (jensj12) [inline MeesGame/UI/DefaultUIValues.settings]: Merge mistake

File: MeesGame/Gameplay/GameStates/LoadMenuState.cs [hunk 8]
```diff
@@ -64,10 +65,18 @@ public LoadMenuState()
                 BeginLevel();
             });
 
+            editorButton = new SpriteSheetButton(new CenteredLocation(0, 25, true, false), null, Strings.map_editor, (UIComponent o) =>
+            {
+                GameEnvironment.GameStateManager.SwitchTo("EditorState");
+                //reset the gamestate to open blank;
+                GameEnvironment.GameStateManager.GetGameState("EditorState").Reset();
```
Comment (jensj12) [inline MeesGame/Gameplay/GameStates/LoadMenuState.cs]: Use CurrentGameState, there's a property for that.

**Comment by jensj12** [comment]
Main menu and quit game buttons in settings menu don't work.

**Comment by twjvdhorst** [comment]
Weird, they worked just fine before... I'll see what I can do

**Review by jensj12 — APPROVED** [review]

**Comment by jensj12** [comment]
MathHelper.Clamp works just fine.
I'd like to see the edit level button renamed to 'Level editor' and moved to the bottom of the screen. I've merged it for now so others can continue upon this as well. (create a new branch if you want to continue working on this)

**Pull request merged**